### PR TITLE
Allow adopters to issue auth tokens to enable search

### PR DESCRIPTION
**Pull request by erikxiv**
## Hey, I just made a Pull Request!

Workaround PR to fix https://github.com/backstage/backstage/issues/5207 while waiting for a more permanent solution to https://github.com/backstage/backstage/issues/5210.

The new search functionality indexes the catalog in a background process which fails for adopters applying auth to their APIs using the outline in https://github.com/backstage/backstage/blob/master/contrib/docs/tutorials/authenticate-api-requests.md. To make this work, this PR allows the search indexer to issue auth tokens.

* Extend the experimental `IdentityClient` in the `auth-backend` plugin to issue tokens. Since plugins may be deployed separately, the dependency between IdentityClient and the `auth-backend` plugin is changed from an API request to retrieve public keys to a database dependency to allow other plugins to generate keys for token issuing and store the public counterpart in the shared database. This is a breaking change for those using api auth as the IdentityClient needs to be instantiated slightly differently.
* ~~Extend DefaultCatalogCollator to optionally use a provided IdentityClient to authorize requests with self-issued tokens~~
* Update relevant contrib doc

Some notable parts:
* ~~Might be cleaner to have adopters make a copy of the DefaultCatalogCollator instead of polluting the default one (changed)~~
* Allowing search plugin to access the auth plugin database breaks the plugin pattern
* Some token parameters (issuer and key duration) are duplicated. Might benefit from moving to config instead.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
Will take a proper review of this once @Rugvip and @freben are back from there holidays in the next 2 weeks :) 🏖️

File: plugins/auth-backend/src/identity/IdentityClient.ts [hunk 1]
```diff
@@ -28,16 +34,37 @@ const CLOCK_MARGIN_S = 10;
...
+    discovery: PluginEndpointDiscovery;
+    logger: Logger;
+  }): Promise<IdentityClient> {
+    const issuer = await options.discovery.getExternalBaseUrl('auth');
+    return new IdentityClient(
+      issuer,
+      new TokenFactory({
+        issuer: issuer,
+        logger: options.logger.child({ component: 'token-factory' }),
+        keyStore: await DatabaseKeyStore.create({
```
Comment (Rugvip) [inline plugins/auth-backend/src/identity/IdentityClient.ts:48]: I'm extremely hesitant towards this to be honest. It's not good to have a collection of different microservices using the same database, which is what we end up with here. What do you think about exploring a more direct shared secret approach to service-to-service auth that operates separately from the OpenID-like approach that we use for the user tokens?

**Review by Rugvip — COMMENTED** [review]
Back from vacation! 🎉 😅

**Comment by Rugvip** [comment]
We're moving forward with #7253, which will include some form of backend tokens that have their own management that is separate from the user tokens. Do you have any thoughts on that way forward and is it alright to close this?

**Comment by erikxiv** [comment]
Closing this in favor of the new discussions

**Pull request closed without merging**